network-proxy-service/src/sshforwarding.rs

```diff
@@ -0,0 +1,150 @@
+        let pem = decode(&self.config.ssh_private_key_base64)?;
+
+        let key_pair = Arc::new(key::KeyPair::RSA {
+            key: openssl::rsa::Rsa::private_key_from_pem(&pem)?,
+            hash: key::SignatureHash::SHA2_256,
+        });
+
+        let sc = self.ssh_client.as_mut().ok_or(Error::SshClientUnInitialized)?;
+        if !sc.authenticate_publickey(&self.config.ssh_user, key_pair).await? {
+            return Err(Error::InvalidSshCredential)
+        }
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl NetworkProxy for SshForwarding {
+    async fn prepare(&mut self) -> Result<(), Error> {
+        self.prepare_ssh_client().await?;
+        self.prepare_local_listener().await?;
+        self.authenticate().await?;
+        Ok(())
+    }
+
+    async fn start_serve(&mut self) -> Result<(), Error> {
+        let sc = self.ssh_client.as_mut().ok_or(Error::SshClientUnInitialized)?;
+        let ll = self.local_listener.as_mut().ok_or(Error::LocalListenerUnInitialized)?;
+        loop {
+            let (forward_stream, _) = ll.accept().await?;
+            let bastion_channel = sc.channel_open_direct_tcpip(
+                &self.config.remote_host,
+                self.config.remote_port as u32,
+                "127.0.0.1", 0).await?;
+            tokio::task::spawn(async move {
+                tunnel_streaming(forward_stream, bastion_channel).await.or_bail("tunnel_handle failed.");
+            });
+        }
+    }
+}
+
+async fn tunnel_streaming(mut forward_stream: TcpStream, mut bastion_channel: client::Channel) -> Result<(), Error>{
+    let mut buf_forward_stream = vec![0; 2048];
+
+    loop {
+        select! {
+            bytes_read = forward_stream.read(&mut buf_forward_stream).fuse() => match bytes_read? {
+                0 => {
+                    bastion_channel.eof().await?;
+                    break
+                },
+                n => {
+                    bastion_channel.data(&buf_forward_stream[..n]).await?;
+                },
+            },
+            bastion_channel_data = bastion_channel.wait().fuse() => match bastion_channel_data {
+                None => {},
+                Some(chan_data) => match chan_data {
+                    thrussh::ChannelMsg::Eof => {
+                      forward_stream.flush().await?;
+                      break;
+                    },
+
+                    thrussh::ChannelMsg::Data { ref data } => {
+                        forward_stream.write(&data).await?;
+                    },
+                    _ => {}
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+pub struct ClientHandler {}
+
+impl client::Handler for ClientHandler {
+    type Error = thrussh::Error;
+    type FutureUnit = futures::future::Ready<Result<(Self, client::Session), Self::Error>>;
+    type FutureBool = futures::future::Ready<Result<(Self, bool), Self::Error>>;
+
+    fn finished_bool(self, b: bool) -> Self::FutureBool {
```

Comment:
I am not exactly sure if I can well answer this.

My understanding is that the API is designed to handle multiple use cases. According to the [doc](https://docs.rs/thrussh/latest/thrussh/):
```
...clients are generally used both in a synchronous way (in the case of SSH, we can think of sending a shell command), and asynchronously (because the server may send unsollicited messages), and hence need to handle multiple interfaces.
``` 
The tunneling use case is simple, so the immediate ready values are good enough. I guess more complex use cases, maybe e.g. involving user actions in the loop, might require non-trivial futures to be returned. However I did not find good examples for that.